pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -862,105 +862,137 @@ func TestPriorityQueue_Pop(t *testing.T) {
+		t.Run(tt.name, func(t *testing.T) {
+			logger, ctx := ktesting.NewTestContext(t)
+			objs := []runtime.Object{highPriorityPodInfo.Pod, unschedulablePodInfo.Pod, medPriorityPodInfo.Pod}
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			q := NewTestQueueWithObjects(ctx, newDefaultQueueSort(), objs, WithClock(c))
 
-	// To simulate the pod is failed in scheduling in the real world, Pop() the pod from activeQ before testing AddUnschedulableIfNotPresent.
-	err = q.activeQ.Add(podInfo)
-	if err != nil {
-		t.Fatalf("unexpected error from activeQ.Add: %v", err)
-	}
-	if p, err := q.Pop(logger); err != nil || p.Pod != medPriorityPodInfo.Pod {
-		t.Errorf("Expected: %v after Pop, but got: %v", medPriorityPodInfo.Pod.Name, p.Pod.Name)
-	}
-	// updating a pod which is in unschedulable queue, and it is not backing off,
-	// we will move it to active queue
-	err = q.AddUnschedulableIfNotPresent(logger, q.newQueuedPodInfo(medPriorityPodInfo.Pod, "plugin"), q.SchedulingCycle())
-	if err != nil {
-		t.Fatalf("unexpected error from AddUnschedulableIfNotPresent: %v", err)
-	}
-	if len(q.unschedulablePods.podInfoMap) != 1 {
-		t.Error("Expected unschedulablePods to be 1.")
-	}
-	updatedPod = medPriorityPodInfo.Pod.DeepCopy()
-	updatedPod.Annotations["foo"] = "test1"
-	// Move clock by podInitialBackoffDuration, so that pods in the unschedulablePods would pass the backing off,
-	// and the pods will be moved into activeQ.
-	c.Step(q.podInitialBackoffDuration)
-	q.Update(logger, medPriorityPodInfo.Pod, updatedPod)
-	if p, err := q.Pop(logger); err != nil || p.Pod != updatedPod {
-		t.Errorf("Expected: %v after Pop, but got: %v", updatedPod.Name, p.Pod.Name)
+			oldPod, newPod, wantUpdatedPod := tt.prepareFunc(logger, q)
+
+			if err := q.Update(logger, oldPod, newPod); err != nil {
+				t.Fatalf("unexpected error from Update: %v", err)
+			}
+
+			var pInfo *framework.QueuedPodInfo
+
+			// validate expected queue
+			if obj, exists, _ := q.podBackoffQ.Get(newQueuedPodInfoForLookup(newPod)); exists {
+				if tt.expectedQ != backoffQ {
+					t.Errorf("expected pod %s not to be queued to backoffQ, but it was", newPod.Name)
+				}
+				pInfo = obj.(*framework.QueuedPodInfo)
+			}
+
+			if obj, exists, _ := q.activeQ.Get(newQueuedPodInfoForLookup(newPod)); exists {
+				if tt.expectedQ != activeQ {
+					t.Errorf("expected pod %s not to be queued to activeQ, but it was", newPod.Name)
+				}
+				pInfo = obj.(*framework.QueuedPodInfo)
+			}
+
+			if pInfoFromUnsched := q.unschedulablePods.get(newPod); pInfoFromUnsched != nil {
+				if tt.expectedQ != unschedulablePods {
+					t.Errorf("expected pod %s to not be queued to unschedulablePods, but it was", newPod.Name)
+				}
+				pInfo = pInfoFromUnsched
+			}
+
+			if diff := cmp.Diff(wantUpdatedPod, pInfo, cmpopts.IgnoreFields(framework.QueuedPodInfo{}, "Timestamp", "InitialAttemptTimestamp")); diff != "" {
+				t.Errorf("Unexpected updated pod diff (-want, +got): %s", diff)
+			}
```

Comment:
Is this comparison that useful? What in the `QueuedPodInfo` is important to check?
Can we skip the fields that are not interesting so that we can build the wantQueuedPodInfo without the need of a function?
I don't particularly like the idea of `prepare` as it makes the test quite hard to read.

Reply:
Ideally, this comparison is to check if `update` action takes effect.
In this test, all the update action we do is for pod's annotations
```
				updatedPod := medPriorityPodInfo.Pod.DeepCopy()
				updatedPod.Annotations["foo"] = "test"
```
So I think maybe we can just check the newpod with the pod we got from queue?
```
if diff := cmp.Diff(newPod, pInfo.PodInfo.Pod); diff != ""
```